File: agents/tools/builtin/grep_tool.py
"""
Grep Tool — search file contents with regex support.

Falls back to Python's re module (no external dependencies).
"""

from __future__ import annotations

import os
import re
from typing import Any

from pydantic import BaseModel, Field

from agents.tools.registry import ToolDefinition, ToolResult, ToolContext
from agents.tools.builtin.path_safety import resolve_path


class GrepInput(BaseModel):
    """Input schema for the grep tool."""
    pattern: str = Field(description="The regex pattern to search for.")
    path: str = Field(description="File or directory path to search in.")
    include: str = Field(default="", description="Glob pattern to filter files (e.g. '*.py'). Only used when path is a directory.")
    case_insensitive: bool = Field(default=False, description="If true, performs case-insensitive search.")
    max_results: int = Field(default=50, description="Maximum number of matches to return.")


async def grep_execute(input_data: dict[str, Any], context: ToolContext) -> ToolResult:
    """Search for a pattern in files."""
    pattern = input_data["pattern"]
    path = input_data["path"]
    include = input_data.get("include", "")
    case_insensitive = input_data.get("case_insensitive", False)
    max_results = input_data.get("max_results", 50)

    try:
        path = resolve_path(path, context)
    except ValueError as exc:
        return ToolResult(data=str(exc), is_error=True)

    if not os.path.exists(path):
        return ToolResult(data=f"Path not found: {path}", is_error=True)

    flags = re.IGNORECASE if case_insensitive else 0
    try:
        compiled = re.compile(pattern, flags)
    except re.error as exc:
        return ToolResult(data=f"Invalid regex pattern: {exc}", is_error=True)

    matches: list[str] = []

    def search_file(filepath: str) -> None:
        """Search a single file for the pattern."""
        try:
            with open(filepath, "r", encoding="utf-8", errors="replace") as f:
                for line_num, line in enumerate(f, 1):
                    if len(matches) >= max_results:
                        return
                    if compiled.search(line):
                        rel_path = os.path.relpath(filepath, context.workspace_dir or os.getcwd())
                        matches.append(f"{rel_path}:{line_num}: {line.rstrip()}")
        except (OSError, UnicodeDecodeError):
            pass  # Skip binary/unreadable files

    if os.path.isfile(path):
        search_file(path)
    else:
        # Walk directory
        import fnmatch
        for root, dirs, files in os.walk(path):
            # Skip common non-useful directories
            dirs[:] = [d for d in dirs if d not in {
                ".git", "__pycache__", "node_modules", ".next",
                ".venv", "venv", ".mypy_cache", ".pytest_cache",
            }]
            for filename in sorted(files):
                if len(matches) >= max_results:
                    break
                if include and not fnmatch.fnmatch(filename, include):
                    continue
                filepath = os.path.join(root, filename)
                search_file(filepath)

    if not matches:
        return ToolResult(data=f"No matches found for pattern: {pattern}")

    result_text = "\n".join(matches)
    header = f"Found {len(matches)} match(es)"
    if len(matches) >= max_results:
        header += f" (capped at {max_results})"

    return ToolResult(data=f"{header}:\n\n{result_text}")


grep_tool = ToolDefinition(
    name="grep",
    description=(
        "Search for a regex pattern in file(s). Can search a single file "
        "or recursively search a directory. Returns matching lines with "
        "file paths and line numbers."
    ),
    input_schema=GrepInput,
    execute=grep_execute,
)

File: cli/src/tools/GrepTool.ts
// tools/GrepTool.ts — regex search across files using Node.js (ripgrep when available)
import { readFile, readdir, stat } from 'fs/promises'
import { join, resolve, relative, extname } from 'path'
import type { RegisteredTool } from '../agent/ToolRegistry'

const EXCLUDE_DIRS = new Set(['node_modules', '.git', '.next', 'dist', 'build', '__pycache__', '.venv'])
const MAX_RESULTS = 250

async function* walk(dir: string): AsyncGenerator<string> {
  let entries
  try {
    entries = await readdir(dir, { withFileTypes: true })
  } catch {
    return
  }
  for (const entry of entries) {
    if (EXCLUDE_DIRS.has(entry.name)) continue
    const full = join(dir, entry.name)
    if (entry.isDirectory()) {
      yield* walk(full)
    } else if (entry.isFile()) {
      yield full
    }
  }
}

function matchesGlob(filename: string, glob: string): boolean {
  if (!glob) return true
  if (glob.startsWith('*.')) {
    return filename.endsWith(glob.slice(1))
  }
  return filename.includes(glob.replace('*', ''))
}

export const GrepTool: RegisteredTool = {
  definition: {
    name: 'grep',
    description: 'Search for a regex pattern across files in a directory. Returns matching file paths and line contents.',
    parameters: {
      type: 'object',
      properties: {
        pattern: { type: 'string', description: 'Regular expression pattern to search for.' },
        path: { type: 'string', description: 'Directory or file to search in. Defaults to current directory.' },
        glob: { type: 'string', description: 'File glob filter (e.g. "*.ts", "*.py"). Optional.' },
        case_insensitive: { type: 'boolean', description: 'Whether to search case-insensitively. Default false.' },
        max_results: { type: 'number', description: `Max results to return. Defaults to ${MAX_RESULTS}.` },
      },
      required: ['pattern'],
    },
  },

  async handler(args, signal) {
    const rawPattern = args['pattern']
    if (typeof rawPattern !== 'string' || !rawPattern) {
      return { output: 'pattern is required.', isError: true }
    }

    const flags = args['case_insensitive'] === true ? 'i' : ''
    let regex: RegExp
    try {
      regex = new RegExp(rawPattern, flags)
    } catch {
      return { output: `Invalid regex: ${rawPattern}`, isError: true }
    }

    const searchPath = typeof args['path'] === 'string' ? resolve(process.cwd(), args['path']) : process.cwd()
    const glob = typeof args['glob'] === 'string' ? args['glob'] : ''
    const maxResults = typeof args['max_results'] === 'number' ? args['max_results'] : MAX_RESULTS

    const results: string[] = []

    try {
      const info = await stat(searchPath)
      const filesToSearch: string[] = info.isFile() ? [searchPath] : []

      if (!info.isFile()) {
        for await (const file of walk(searchPath)) {
          if (signal?.aborted) break
          if (glob && !matchesGlob(file, glob)) continue
          filesToSearch.push(file)
        }
      }

      for (const file of filesToSearch) {
        if (signal?.aborted) break
        if (results.length >= maxResults) break

        let text: string
        try {
          text = await readFile(file, 'utf-8')
        } catch {
          continue // Skip binary or unreadable files
        }

        const lines = text.split('\n')
        const relFile = relative(process.cwd(), file)

        for (let i = 0; i < lines.length; i++) {
          if (results.length >= maxResults) break
          if (regex.test(lines[i]!)) {
            results.push(`${relFile}:${i + 1}: ${lines[i]}`)
          }
        }
      }
    } catch (err: unknown) {
      const msg = err instanceof Error ? err.message : String(err)
      return { output: `Grep error: ${msg}`, isError: true }
    }

    if (results.length === 0) {
      return { output: 'No matches found.', isError: false }
    }

    const truncated = results.length >= maxResults
    const header = `Found ${results.length} match${results.length === 1 ? '' : 'es'}${truncated ? ` (showing first ${maxResults})` : ''}:`
    return { output: `${header}\n${results.join('\n')}`, isError: false }
  },
}
